sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadTableSource.java

```diff
@@ -234,24 +147,53 @@ public synchronized long getEstimatedSizeBytes(PipelineOptions options)
   }
 
   /**
-   * You can potentially create multiple readers on a source.
+   * Gets the estimated number of bytes returned by the current source. For sources created by a
+   * split() operation, this is a fraction of the total table size.
    */
+  @Override
+  public synchronized long getEstimatedSizeBytes(PipelineOptions options)
+      throws IOException, InterruptedException {
+    if (cachedReadSizeBytes == null) {
+      BigQueryOptions bigQueryOptions = options.as(BigQueryOptions.class);
+      TableReference tableReference = getEffectiveTableReference(bigQueryOptions);
+      Table table = bqServices.getDatasetService(bigQueryOptions).getTable(tableReference);
+      if (table != null) {
+        cachedReadSizeBytes = table.getNumBytes();
+      }
+      if (cachedReadSizeBytes == null) {
+        cachedReadSizeBytes = 0L;
+      }
+    }
+
+    return cachedReadSizeBytes;
+  }
+
   @Override
   public BoundedReader<T> createReader(PipelineOptions options) {
-    return new BigQueryParallelReader<>(
-        readSession,
-        readLocation,
-        (readSessionOptions != null && readSessionOptions.getRowBatchSize() != null)
-            ? readSessionOptions.getRowBatchSize()
-            : 1000,
-        parseFn,
-        this,
-        bqServices,
-        options.as(BigQueryOptions.class));
+    throw new UnsupportedOperationException("BigQuery source must be split before being read");
```

Comment:
This will fail for DirectRunner that does not perform splitting I believe. Also, this will fail if the source is used as a side input.

Reply:
This is what the existing BigQueryTableSource and BigQueryQuerySource do (right?), and also I'm not sure the existing code will work before splitting since the readSession won't be initialized.